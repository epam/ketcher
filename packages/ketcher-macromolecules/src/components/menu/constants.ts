export const SELECT_SUBMENU_ID = 'select-submenu';
