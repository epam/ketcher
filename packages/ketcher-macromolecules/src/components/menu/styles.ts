/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';

export const MenuLayout = styled.div<{ isHorizontal?: boolean }>(
  ({ theme, isHorizontal }) => ({
    backgroundColor: theme.ketcher.color.background.primary,
    borderRadius: '4px',
    display: 'flex',
    flexDirection: isHorizontal ? 'row' : 'column',
  }),
);

export const Divider = styled.hr`
  width: 18px;
  margin: 6px 0;
  align-self: center;
  border-width: thin 0px 0px 0px;
  border-style: solid;
  border-color: ${({ theme }) => theme.ketcher.color.divider};
`;

export const VerticalDivider = styled.hr`
  height: 18px;
  margin: 0px 6px;
  align-self: center;
  border-width: 0px thin 0px 0px;
  border-style: solid;
  border-color: ${({ theme }) => theme.ketcher.color.divider};
`;

export const StyledGroup = styled.div<{ isHorizontal?: boolean }>(
  ({ theme, isHorizontal }) => ({
    display: 'flex',
    flexDirection: isHorizontal ? 'row' : 'column',
    flexWrap: 'nowrap',
    alignItems: 'center',
    backgroundColor: theme.ketcher.color.background.primary,
    borderRadius: '2px',
    width: isHorizontal ? undefined : '32px',
    marginBottom: isHorizontal ? undefined : '8px',
    '> :last-child': isHorizontal
      ? undefined
      : {
          marginBottom: 0,
        },
  }),
);
