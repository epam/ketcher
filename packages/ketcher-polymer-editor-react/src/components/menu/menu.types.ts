/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

type MenuItemVariant =
  | 'open'
  | 'select-lasso'
  | 'select-rectangle'
  | 'select-fragment'
  | 'rectangle'
  | 'ellipse'
  | 'help'
  | 'settings'
  | 'undo'
  | 'single-bond'
  | 'erase'
  | 'rotate'
  | 'horizontal-flip'
  | 'vertical-flip'
  | 'bracket'

export type { MenuItemVariant }
