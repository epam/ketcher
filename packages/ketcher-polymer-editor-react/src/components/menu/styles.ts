/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';

export const MenuLayout = styled.div`
  background-color: ${({ theme }) => theme.ketcher.color.background.primary};
`;

export const Divider = styled.span`
  display: block;
  height: 8px;
  width: 32px;
  border-top: 1px solid;
  border-color: ${({ theme }) => theme.ketcher.color.divider};
`;

export const StyledGroup = styled.div`
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  align-items: center;
  background-color: ${({ theme }) => theme.ketcher.color.background.primary};
  border-radius: 2px;
  width: 32px;
  margin-bottom: 8px;

  > * {
    margin-bottom: 8px;
  }

  > :last-child {
    margin-bottom: 0;
  }
`;
