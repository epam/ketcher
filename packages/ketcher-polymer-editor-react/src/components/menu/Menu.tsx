/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import React from 'react';
import { MenuItem } from './menuItem';
import { SubMenu } from './subMenu';
import { IMenuContext, MenuContext } from '../../contexts';
import { Divider, MenuLayout, StyledGroup } from './styles';
import { GroupProps, MenuProps } from './types';

const Group = ({
  children,
  divider = false,
}: React.PropsWithChildren<GroupProps>) => {
  const subComponents = React.Children.map(
    children as JSX.Element[],
    (child) => {
      return child.type === MenuItem || child.type === SubMenu ? child : null;
    },
  );

  return (
    <>
      <StyledGroup>{subComponents.map((component) => component)}</StyledGroup>
      {divider && <Divider />}
    </>
  );
};

const Menu = ({
  children,
  onItemClick,
  activeMenuItem,
}: React.PropsWithChildren<MenuProps>) => {
  const context = React.useMemo<IMenuContext>(
    () => ({
      isActive: (itemKey) => activeMenuItem === itemKey,
      activate: (itemKey) => {
        onItemClick(itemKey);
      },
    }),
    [activeMenuItem, onItemClick],
  );

  const subComponents = React.Children.map(
    children as JSX.Element[],
    (child) => {
      return child.type === Group ? child : null;
    },
  );

  return (
    <MenuContext.Provider value={context}>
      <MenuLayout>{subComponents.map((component) => component)}</MenuLayout>
    </MenuContext.Provider>
  );
};

Menu.Group = Group;
Menu.Item = MenuItem;
Menu.Submenu = SubMenu;

export { Menu, MenuContext };
