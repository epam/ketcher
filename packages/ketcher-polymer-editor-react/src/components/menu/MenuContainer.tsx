/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { useAppDispatch, useAppSelector } from 'hooks'
import { selectEditorActiveTool, selectTool } from 'state/common'
import { Menu } from 'components/menu/Menu'

const MenuContainer = () => {
  const dispatch = useAppDispatch()
  const activeTool = useAppSelector(selectEditorActiveTool)

  const menuProps = {
    activeItem: activeTool,
    onClick: (tool) => dispatch(selectTool(tool))
  }
  return (
    <Menu>
      <Menu.Group>
        <Menu.Item name="open" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="undo" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="erase" {...menuProps} />
        <Menu.Submenu vertical {...menuProps}>
          <Menu.Item name="select-lasso" {...menuProps} />
          <Menu.Item name="select-rectangle" {...menuProps} />
          <Menu.Item name="select-fragment" {...menuProps} />
        </Menu.Submenu>
        <Menu.Submenu {...menuProps}>
          <Menu.Item name="rectangle" {...menuProps} />
          <Menu.Item name="ellipse" {...menuProps} />
        </Menu.Submenu>
        <Menu.Submenu {...menuProps}>
          <Menu.Item name="rotate" {...menuProps} />
          <Menu.Item name="horizontal-flip" {...menuProps} />
          <Menu.Item name="vertical-flip" {...menuProps} />
        </Menu.Submenu>
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="single-bond" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="bracket" {...menuProps} />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item name="settings" {...menuProps} />
        <Menu.Item name="help" {...menuProps} />
      </Menu.Group>
    </Menu>
  )
}
export { MenuContainer }
