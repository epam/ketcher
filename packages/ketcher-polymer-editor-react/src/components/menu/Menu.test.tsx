/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { render } from '@testing-library/react';
import { Menu } from 'components/menu/Menu';

const menuItemChanged = jest.fn();

const MenuContainer = () => {
  return (
    <Menu onItemClick={menuItemChanged}>
      <Menu.Group>
        <Menu.Item itemId="open" />
      </Menu.Group>
      <Menu.Group>
        <Menu.Item itemId="erase" />
        <Menu.Submenu vertical>
          <Menu.Item itemId="select-lasso" />
          <Menu.Item itemId="select-rectangle" />
          <Menu.Item itemId="select-fragment" />
        </Menu.Submenu>
        <Menu.Submenu>
          <Menu.Item itemId="shape-rectangle" />
          <Menu.Item itemId="shape-ellipse" />
        </Menu.Submenu>
        <Menu.Submenu>
          <Menu.Item itemId="transform-flip-h" />
          <Menu.Item itemId="transform-flip-v" />
        </Menu.Submenu>
      </Menu.Group>
    </Menu>
  );
};

describe('Menu component', () => {
  it('should render menu component in a container', () => {
    expect(render(withThemeProvider(<MenuContainer />))).toMatchSnapshot();
  });
});
