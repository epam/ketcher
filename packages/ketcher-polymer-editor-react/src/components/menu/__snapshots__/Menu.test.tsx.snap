// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component should render menu component in a container 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-10qeqvw"
      >
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
          role="menuitem"
          tabindex="-1"
        >
          <icon-open.svg
            fill="#333333"
            role="img"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </div>
      <div
        class="css-10qeqvw"
      >
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
          role="menuitem"
          tabindex="-1"
        >
          <icon-erase.svg
            fill="#333333"
            role="img"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <div
          class="css-12wx08y"
        >
          <div
            class="css-5su45f"
          >
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
              role="menuitem"
              tabindex="-1"
            >
              <icon-select-lasso.svg
                fill="#333333"
                role="img"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
            <div
              class="css-clax1l"
            >
              <icon-dropdown.svg
                fill="#333333"
                role="img"
              />
            </div>
          </div>
        </div>
        <div
          class="css-12wx08y"
        >
          <div
            class="css-5su45f"
          >
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
              role="menuitem"
              tabindex="-1"
            >
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
            <div
              class="css-clax1l"
            >
              <icon-dropdown.svg
                fill="#333333"
                role="img"
              />
            </div>
          </div>
        </div>
        <div
          class="css-12wx08y"
        >
          <div
            class="css-5su45f"
          >
            <li
              class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
              role="menuitem"
              tabindex="-1"
            >
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </li>
            <div
              class="css-clax1l"
            >
              <icon-dropdown.svg
                fill="#333333"
                role="img"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-10qeqvw"
    >
      <li
        class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
        role="menuitem"
        tabindex="-1"
      >
        <icon-open.svg
          fill="#333333"
          role="img"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </li>
    </div>
    <div
      class="css-10qeqvw"
    >
      <li
        class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
        role="menuitem"
        tabindex="-1"
      >
        <icon-erase.svg
          fill="#333333"
          role="img"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </li>
      <div
        class="css-12wx08y"
      >
        <div
          class="css-5su45f"
        >
          <li
            class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
            role="menuitem"
            tabindex="-1"
          >
            <icon-select-lasso.svg
              fill="#333333"
              role="img"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <div
            class="css-clax1l"
          >
            <icon-dropdown.svg
              fill="#333333"
              role="img"
            />
          </div>
        </div>
      </div>
      <div
        class="css-12wx08y"
      >
        <div
          class="css-5su45f"
        >
          <li
            class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <div
            class="css-clax1l"
          >
            <icon-dropdown.svg
              fill="#333333"
              role="img"
            />
          </div>
        </div>
      </div>
      <div
        class="css-12wx08y"
      >
        <div
          class="css-5su45f"
        >
          <li
            class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-sxwttn-MuiButtonBase-root-MuiMenuItem-root-StyledMenuButton"
            role="menuitem"
            tabindex="-1"
          >
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </li>
          <div
            class="css-clax1l"
          >
            <icon-dropdown.svg
              fill="#333333"
              role="img"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
